lib/msf/core/post/file.rb

```diff
@@ -248,11 +248,12 @@ def executable?(path)
   def writable?(path)
     verification_token = Rex::Text.rand_text_alpha_upper(8)
     if session.type == 'powershell' && file?(path)
-      return cmd_exec("$a=[System.IO.File]::OpenWrite('#{path}');if($?){echo #{verification_token}};$a.Close()").include?(verification_token)
+      cmd_exec("$a=[System.IO.File]::OpenWrite('#{path}');if($?){echo #{verification_token}};$a.Close()").include?(verification_token)
+    elsif session.platform == 'windows'
+      cmd_exec("echo #{Rex::Text.rand_text_alphanumeric(5..30)} >> \"#{path}\" && echo true").to_s.include? 'true'
```

Comment:
I think a `writable?` function should avoid writing to the file, as it may cause unintended side effects

Potentially this could use `icacls "#{path}"`, with some additional parsing? 🤔

Reply:
The problem with `icacls` that I find is that inheritance of permissions on Windows is a complex beast and whilst we would be able to get the inherited permissions via `icacls` output, parsing it is a little more complex as it would only show us groups.

Example:

```
C:\Users\*redacted*>icacls C:\Windows\Synaptics.log
C:\Windows\Synaptics.log NT AUTHORITY\SYSTEM:(I)(F)
                         BUILTIN\Administrators:(I)(F)
                         BUILTIN\Users:(I)(RX)
                         APPLICATION PACKAGE AUTHORITY\ALL APPLICATION PACKAGES:(I)(RX)
                         APPLICATION PACKAGE AUTHORITY\ALL RESTRICTED APPLICATION PACKAGES:(I)(RX)

Successfully processed 1 files; Failed processing 0 files
```
That output is great, but then we don't know if the current user is part of those groups or not. Is it possible though? Yes probably but the user could be part of multiple groups with different permissions set on them which complicates matters further.

I'd be down for doing this if we had a way to approach this problem but right now it seems like there are a number of concerns that would have to be addressed.

I do agree however that we shouldn't be writing to the file as this could have unintended side effects.

Reply:
Hmm might be possible to do this with `whoami /groups` and with `icacls`. Only issue I see potentially is that `whoami /groups` will also output SSIDs for unknown groups and I can't see a way to get this with `icacls`.

Reply:
@adfoster-r7 New approach taken in https://github.com/rapid7/metasploit-framework/pull/15987/commits/a204591f98006a89996bc105b6f15a5680bad099, thanks for the advice on this!